gloo/transport/tcp/tls/pair.cc

```diff
@@ -0,0 +1,284 @@
+  std::array<struct iovec, 2> iov;
+  int ioc;
+
+  const auto opcode = op.getOpcode();
+
+  // Acquire pointer to unbound buffer if applicable.
+  if (opcode == Op::SEND_UNBOUND_BUFFER) {
+    buf = NonOwningPtr<UnboundBuffer>(op.ubuf);
+    if (!buf) {
+      return false;
+    }
+  }
+
+  const auto nbytes = prepareWrite(op, buf, iov.data(), ioc);
+  ssize_t total_rv = 0;
+  for (int i = 0; i < ioc; ++i) {
+    for (;;) {
+      if (iov[i].iov_len == 0) {
+        break;
+      }
+      ssize_t rv = SSL_write(ssl_, iov[i].iov_base, iov[i].iov_len);
+      if (rv <= 0) {
+        int err = SSL_get_error(ssl_, rv);
+
+        GLOO_ENFORCE(err != SSL_ERROR_NONE);
+        GLOO_ENFORCE(err != SSL_ERROR_WANT_READ);
+
+        if (err == SSL_ERROR_WANT_WRITE) {
+          // just repeat the same write
+          continue;
+        }
+
+        if (err == SSL_ERROR_SYSCALL) {
+          fatal_error_occurred_ = true;
+          if (errno == EPIPE) {
+            if (!sync_) {
+              return false;
+            }
+          }
+        }
+
+        // Unexpected error
+        signalException(GLOO_ERROR_MSG("SSL_write ", peer_.str(), " failed: ",
+                                       ", errno = ", strerror(errno),
+                                       ", ssl error: ", getSSLErrorMessage()));
+        return false;
+      }
+      total_rv += rv;
+      op.nwritten += rv;
+      break;
+    }
+  }
+  GLOO_ENFORCE_EQ(total_rv, nbytes);
+  GLOO_ENFORCE_EQ(op.nwritten, op.preamble.nbytes);
+
+  writeComplete(op, buf, opcode);
+  return true;
+}
+
+bool Pair::read() {
+  NonOwningPtr<UnboundBuffer> buf;
+  auto start = std::chrono::steady_clock::now();
+
+  for (;;) {
+    struct iovec iov = {
+        .iov_base = nullptr,
+        .iov_len = 0,
+    };
+    const auto nbytes = prepareRead(rx_, buf, iov);
+    if (nbytes < 0) {
+      return false;
+    }
+
+    // Break from loop if the op is complete.
+    // Note that this means that the buffer pointer has been
+    // set, per the call to prepareRead.
+    if (nbytes == 0) {
+      break;
+    }
+
+    ssize_t rv = 0;
+    for (;;) {
+      rv = SSL_read(ssl_, iov.iov_base, iov.iov_len);
+      if (rv <= 0) {
+        int err = SSL_get_error(ssl_, rv);
+
+        GLOO_ENFORCE(err != SSL_ERROR_NONE);
+        GLOO_ENFORCE(err != SSL_ERROR_WANT_WRITE);
+
+        if (err == SSL_ERROR_WANT_READ) {
+          return false;
+        }
+
+        if (err == SSL_ERROR_ZERO_RETURN) {
+          if (errno == ECONNRESET) {
+            if (!sync_) {
+              return false;
+            }
+          }
+        }
+
+        if (err == SSL_ERROR_SYSCALL) {
+          fatal_error_occurred_ = true;
+          if (errno == EPIPE) {
+            if (!sync_) {
+              return false;
+            }
+          }
+          if (errno == ECONNRESET) {
+            if (!sync_) {
+              return false; // Exit if Connection reset by peer on read
+            }
+          }
```

Comment:
any reason these are add as two different branches? will `errno == EPIPE  || errno == ECONNRESET` be sufficient?

Reply:
addressed